prow/cmd/build/controller.go

```diff
@@ -17,46 +17,423 @@ limitations under the License.
-	_ "k8s.io/client-go/tools/cache"
-	_ "k8s.io/client-go/tools/clientcmd"
-	_ "k8s.io/client-go/tools/record"
-	_ "k8s.io/client-go/util/workqueue"
+	"errors"
+	"flag"
+	"fmt"
+	"os"
+	"os/signal"
+	"reflect"
+	"syscall"
+	"time"
+
+	prowjobv1 "k8s.io/test-infra/prow/apis/prowjobs/v1"
+	prowjobset "k8s.io/test-infra/prow/client/clientset/versioned"
+	prowjobscheme "k8s.io/test-infra/prow/client/clientset/versioned/scheme"
+	prowjobinfo "k8s.io/test-infra/prow/client/informers/externalversions"
+	prowjobinfov1 "k8s.io/test-infra/prow/client/informers/externalversions/prowjobs/v1"
+	prowjoblisters "k8s.io/test-infra/prow/client/listers/prowjobs/v1"
+	"k8s.io/test-infra/prow/logrusutil"
+
+	buildv1alpha1 "github.com/knative/build/pkg/apis/build/v1alpha1"
+	buildset "github.com/knative/build/pkg/client/clientset/versioned"
+	buildinfo "github.com/knative/build/pkg/client/informers/externalversions"
+	buildinfov1alpha1 "github.com/knative/build/pkg/client/informers/externalversions/build/v1alpha1"
+	buildlisters "github.com/knative/build/pkg/client/listers/build/v1alpha1"
+
+	untypedcorev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/workqueue"
 
 	"github.com/sirupsen/logrus"
 )
 
+var (
+	masterURL      string
+	kubeconfig     string
+	controllerName = "prow-build-crd"
+)
+
+func init() {
+	flag.StringVar(&kubeconfig, "kubeconfig", "", "Path to kubeconfig. Only required if out of cluster")
+	flag.StringVar(&masterURL, "master", "", "The address of the kubernetes API server. Overrides any value in kubeconfig. Only required if out of cluster")
+}
+
+func stopper() chan struct{} {
+	stop := make(chan struct{})
+	c := make(chan os.Signal, 2)
+	signal.Notify(c, os.Interrupt, syscall.SIGTERM)
+	go func() {
+		<-c
+		close(stop)
+		<-c
+		os.Exit(1)
+	}()
+	return stop
+}
+
 func main() {
-	logrus.Fatalf("see next pr")
+	flag.Parse()
+	logrusutil.NewDefaultFieldsFormatter(nil, logrus.Fields{"component": "build"})
+	stop := stopper()
+
+	cfg, err := clientcmd.BuildConfigFromFlags(masterURL, kubeconfig)
+	if err != nil {
+		logrus.Fatalf("Error building kubeconfig: %v", err)
+	}
+
+	kc, err := kubernetes.NewForConfig(cfg)
+	if err != nil {
+		logrus.Fatalf("Error building kubernetes client: %v", err)
+	}
+
+	pjc, err := prowjobset.NewForConfig(cfg)
+	if err != nil {
+		logrus.Fatalf("Error building prowjob client: %v", err)
+	}
+
+	bc, err := buildset.NewForConfig(cfg)
+	if err != nil {
+		logrus.Fatalf("Error building build client: %v", err)
+	}
+
+	pjif := prowjobinfo.NewSharedInformerFactory(pjc, time.Second*30)
```

Comment:
Is this a 30s relist interval? Probably too tight.

Reply:
Isn't this just for when watches fail? In which case maybe this value is too aggressive?

AKA why do we need to relist objects that have not changed every 30s?

Reply:
Yeah, watches obsolete listing, eg. in core controllers the default relist interval has been pushed to a random value between 12-24h (last time I checked). In general, watches are supposed to be eventually consistent.

Reply:
Updated to 30m